hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/EncryptionS3ClientFactory.java

```diff
@@ -0,0 +1,224 @@
+    }
+
+    s3Client = super.createS3Client(uri, parameters);
+    s3AsyncClient = super.createS3AsyncClient(uri, parameters);
+
+    return createS3EncryptionClient(parameters.getClientSideEncryptionMaterials());
+  }
+
+  /**
+   * Create async encrypted s3 client.
+   * @param uri S3A file system URI
+   * @param parameters parameter object
+   * @return async encrypted s3 client
+   * @throws IOException IO failures
+   */
+  @Override
+  public S3AsyncClient createS3AsyncClient(URI uri, S3ClientCreationParameters parameters)
+      throws IOException {
+    if (!isEncryptionClientAvailable()) {
+      throw unavailable(uri, ENCRYPTION_CLIENT_CLASSNAME, null,
+          "No encryption client available");
+    }
+    return createS3AsyncEncryptionClient(parameters.getClientSideEncryptionMaterials());
+  }
+
+  /**
+   * Create encrypted s3 client.
+   * @param cseMaterials
+   * @return encrypted s3 client
+   */
+  private S3Client createS3EncryptionClient(final CSEMaterials cseMaterials) {
+    Preconditions.checkArgument(s3AsyncClient !=null,
+        "S3 async client not initialized");
+    Preconditions.checkArgument(s3Client !=null,
+        "S3 client not initialized");
+    S3EncryptionClient.Builder s3EncryptionClientBuilder =
+        S3EncryptionClient.builder().wrappedAsyncClient(s3AsyncClient).wrappedClient(s3Client)
+            // this is required for doing S3 ranged GET calls
+            .enableLegacyUnauthenticatedModes(true)
+            // this is required for backward compatibility with older encryption clients
+            .enableLegacyWrappingAlgorithms(true);
+
+    switch (cseMaterials.getCseKeyType()) {
+    case KMS:
+      s3EncryptionClientBuilder.kmsKeyId(cseMaterials.getKmsKeyId());
+      break;
+    case CUSTOM:
+      Keyring keyring = getKeyringProvider(cseMaterials.getCustomKeyringClassName(),
+          cseMaterials.getConf());
+      CryptographicMaterialsManager cmm =  DefaultCryptoMaterialsManager.builder()
+          .keyring(keyring)
+          .build();
+      s3EncryptionClientBuilder.cryptoMaterialsManager(cmm);
+      break;
+    default:
+      break;
+    }
+
+    return s3EncryptionClientBuilder.build();
+  }
+
+  /**
+   * Create async encrypted s3 client.
+   * @param cseMaterials
+   * @return encrypted async s3 client
+   */
+  private S3AsyncClient createS3AsyncEncryptionClient(final CSEMaterials cseMaterials) {
+    Preconditions.checkArgument(s3AsyncClient !=null,
+        "S3 async client not initialized");
+    S3AsyncEncryptionClient.Builder s3EncryptionAsyncClientBuilder =
+        S3AsyncEncryptionClient.builder().wrappedClient(s3AsyncClient)
+            // this is required for doing S3 ranged GET calls
+            .enableLegacyUnauthenticatedModes(true)
+            // this is required for backward compatibility with older encryption clients
+            .enableLegacyWrappingAlgorithms(true);
+
+    switch (cseMaterials.getCseKeyType()) {
```

Comment:
won't that be a overkill here ?

Reply:
> are you sure we still need this? what happens without it? IIRC, I had to do this because if you enabled multipart, all ranged GETs failed. see https://github.com/apache/hadoop/pull/6164/files#r1352852582 for more info. I'm hoping this should have been fixed by the SDK by team now

I don't see any related changes in the change log (https://github.com/aws/amazon-s3-encryption-client-java/blob/v3.1.3/CHANGELOG.md). Is there any specific integration test you want me to run. AFIK async client is only involved in copy / rename operation. Is `testEncryptionOverRename` and good candidate ?

Reply:
Additionally without this change it still fails with
`Caused by: java.lang.UnsupportedOperationException: Multipart download is not yet supported. Instead use the CRT based S3 client for multipart download.
	at software.amazon.awssdk.services.s3.internal.multipart.MultipartS3AsyncClient.getObject(MultipartS3AsyncClient.java:116)
	at software.amazon.encryption.s3.internal.GetEncryptedObjectPipeline.getObject(GetEncryptedObjectPipeline.java:67)
	at software.amazon.encryption.s3.S3EncryptionClient.getObject(S3EncryptionClient.java:252)
	... 36 more`
even using latest client version 3.1.3

Reply:
> is it possible to dedupe this code for sync and async using generics? See DefaultS3ClientFactory.configureClientBuilder for an example.

ack

Reply:
the Multipart download is not yet supported is a bit annoying. This means we can't enable multipartCopy on the async client when using CSE because of this, which may impact rename performance etc. We don't actually wan't multipart download, but if you enable multipart it doesn't let you make ranged GETs either and fails with exception. 

I don't know how much of an issue this is, I'll sync with you offline.